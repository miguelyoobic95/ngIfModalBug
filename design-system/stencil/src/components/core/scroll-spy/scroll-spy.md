---
name: Scroll Spy
category: Core Components
---

## Attributes

|Name|Type|Default|Description|
|---|---|---|---|
|repeat|`boolean`|`false`|Set to true to emit the events several times|

```yoo-scroll-spy-page.html

            <p>Hola this is a scrollable content in a Slim Scroll</p>
            <yoo-badge text="SLIM" class="danger"></yoo-badge>
            <yoo-badge text="SLIM" class="info"></yoo-badge>
            <yoo-badge text="SLIM" class="success"></yoo-badge>
            <yoo-badge text="SLIM" class="warning"></yoo-badge>
            <yoo-badge text="SLIM" class="accent"></yoo-badge>
            <yoo-badge text="SLIM" class="gradient-danger"></yoo-badge>
            <yoo-badge text="SLIM" class="gradient-info"></yoo-badge>
            <yoo-badge text="SLIM" class="gradient-accent"></yoo-badge>
            <yoo-badge text="SLIM" class="gradient-warning"></yoo-badge>
            <br/>
            <yoo-badge text="SCROLL" class="danger"></yoo-badge>
            <br/>
            <yoo-badge text="SCROLL" class="success"></yoo-badge>
            <br/>
            <yoo-badge text="SLIM" class="success"></yoo-badge>
            <br/>
            <p>This is the enddddddddd.... Well not exactly there is a text block after</p>
             <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
             <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
             <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
             <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
             <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
             <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
             <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
            <h1>Spy</h1>
            <yoo-scroll-spy></yoo-scroll-spy>
            <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
             <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
             <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
```

```yoo-scroll-spy-page.js

    var spy = document.querySelector('yoo-scroll-spy');
    spy.addEventListener('enterInView', () => console.log('enter in view'));
    spy.addEventListener('outOfView', () => console.log('out of view'));
```

```yoo-scroll-spy-yoo-slim-scroll.html

    <yoo-slim-scroll height="200px">
        <div>
            <p>Hola this is a scrollable content in a Slim Scroll</p>
            <yoo-badge text="SLIM" class="danger"></yoo-badge>
            <yoo-badge text="SLIM" class="info"></yoo-badge>
            <yoo-badge text="SLIM" class="success"></yoo-badge>
            <yoo-badge text="SLIM" class="warning"></yoo-badge>
            <yoo-badge text="SLIM" class="accent"></yoo-badge>
            <yoo-badge text="SLIM" class="gradient-danger"></yoo-badge>
            <yoo-badge text="SLIM" class="gradient-info"></yoo-badge>
            <yoo-badge text="SLIM" class="gradient-accent"></yoo-badge>
            <yoo-badge text="SLIM" class="gradient-warning"></yoo-badge>
            <br/>
            <yoo-badge text="SCROLL" class="danger"></yoo-badge>
            <br/>
            <yoo-badge text="SCROLL" class="success"></yoo-badge>
            <br/>
            <yoo-badge text="SLIM" class="success"></yoo-badge>
            <br/>
            <p>This is the enddddddddd.... Well not exactly there is a spy after</p>
            <h1>Spy</h1>
            <yoo-scroll-spy></yoo-scroll-spy>
            <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
            <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
             <p>
            Lorem ipsum dolor amet pork belly man braid PBR&B stumptown prism drinking vinegar lomo neutra organic keffiyeh. Taiyaki VHS pug, snackwave cliche polaroid subway tile jean shorts typewriter small batch. Unicorn synth slow-carb bitters, hashtag retro butcher DIY vape austin cliche hoodie semiotics. Brooklyn disrupt cliche green juice, unicorn meditation flexitarian palo santo pop-up. Quinoa air plant post-ironic 8-bit. 8-bit flexitarian blue bottle, try-hard aesthetic biodiesel forage. Cliche raw denim pop-up, blue bottle sartorial meggings farm-to-table ethical four loko polaroid bicycle rights disrupt af mixtape artisan.
            </p>
          </div>
    </yoo-slim-scroll>
```

```yoo-scroll-spy-yoo-slim-scroll.js

    var spy = document.querySelector('yoo-scroll-spy');
    spy.addEventListener('enterInView', () => console.log('enter in view'));
    spy.addEventListener('outOfView', () => console.log('out of view'));
```

## Methods

|Name|Parameters|Description|
|---|---|---|

## Events

|Attr|Description|
|---|---|---|
|`enterInView`|Emitted when the component become visible|
|`outOfView`|Emitted when the component become invisible|