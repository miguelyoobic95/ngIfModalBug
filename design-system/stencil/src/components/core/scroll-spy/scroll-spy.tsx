import { Component, Prop, Event, EventEmitter, Element, State, Listen } from '@stencil/core';

@Component({
    tag: 'yoo-scroll-spy',
    styleUrl: 'scroll-spy.scss',
    scoped: true
})
export class YooScrollSpyComponent {

    @Prop() repeat: boolean = true;

    @Event() enterInView: EventEmitter<boolean>;
    @Event() outOfView: EventEmitter<boolean>;

    @State() isInView: boolean;
    @State() enterEmitted: boolean;
    @State() outEmitted: boolean;
    @State() parentScroll: any;

     @Element() host: HTMLStencilElement;

    @Listen('body:scroll')
    bodyScrollListener() {
        this.onScroll();
    }

    componentDidLoad() {
        let parentScroll = this.host.closest('yoo-slim-scroll');
        if (parentScroll) {
            this.parentScroll = parentScroll;
            this.parentScroll.addEventListener('scrollStart', () => this.onScroll());
            this.parentScroll.addEventListener('scrollEnd', () => this.onScroll());
        }
        this.isInView = this.isElementInViewport();
        if (this.isInView) {
            this.enterInView.emit(true);
            this.enterEmitted = true;
        }
    }

    onScroll() {
        let newIsInView: boolean = this.isElementInViewport();
        if ((!this.isInView && newIsInView) && ( this.repeat || !this.enterEmitted)) {
            this.enterInView.emit(true);
            this.enterEmitted = true;
        }
        if ((this.isInView && !newIsInView) && (this.repeat || !this.outEmitted)) {
            this.outOfView.emit(true);
            this.outEmitted = true;
        }
        this.isInView = newIsInView;
    }

    isElementInViewport(): boolean {
        let rect = this.host.getBoundingClientRect();
        if (! this.parentScroll) {
            return (rect.top >= 0 &&
                rect.left >= 0 &&
                rect.bottom <= (window.innerHeight || document.documentElement.clientHeight ) &&
                rect.right <= (window.innerWidth || document.documentElement.clientWidth)
            );
        } else {
            let rectScroll = this.parentScroll.getBoundingClientRect();
            return (rect.top >= 0 && rect.top >= rectScroll.top &&
                rect.left >= 0 && rect.left >= rectScroll.left &&
                rect.bottom  <= rectScroll.bottom &&
                rect.right <= rectScroll.right
            );
        }
    }

    render(): JSX.Element {
        return (
            <div></div>
        );
    }
}
