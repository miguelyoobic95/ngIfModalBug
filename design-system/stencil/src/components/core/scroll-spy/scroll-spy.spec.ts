import { TestWindow } from '@stencil/core/testing';

import { YooScrollSpyComponent } from './scroll-spy';

describe('YooScrollSpyComponent', () => {
    it('should build', () => {
        expect(new YooScrollSpyComponent()).toBeTruthy();
    });

    describe('Rendering', () => {
        it ('Should render with a basic structure', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooScrollSpyComponent],
                html: '<yoo-scroll-spy></yoo-scroll-spy>'
            });
            expect(element).toMatchSnapshot();
        });
    });

    describe('events', () => {
        // xit ('should emit on enter in view', async () => {
        //     let window = new TestWindow();
        //     let element = await window.load({
        //         components: [YooScrollSpyComponent],
        //         html: '<yoo-scroll-spy></yoo-scroll-spy>'
        //     });
        //
        //
        //     let res;
        //     let emitItem = jest.fn(ev => {res = ev.detail});
        //     window.document.addEventListener('enterInView', emitItem);
        //
        //     // TODO: HERE call the action to trigger the enter in view
        //
        //     expect(emitItem).toHaveBeenCalled();
        //     expect(res).toEqual(true);
        // });

        // xit ('should emit on out of view', async () => {
        //     let window = new TestWindow();
        //     let element = await window.load({
        //         components: [YooScrollSpyComponent],
        //         html: '<yoo-scroll-spy></yoo-scroll-spy>'
        //     });
        //
        //     let res;
        //     let emitItem = jest.fn(ev => {res = ev.detail});
        //     window.document.addEventListener('outOfView', emitItem);
        //
        //     // TODO: HERE call the action to trigger the out of view
        //
        //     expect(emitItem).toHaveBeenCalled();
        //     expect(res).toEqual(true);
        // });
    });
});
