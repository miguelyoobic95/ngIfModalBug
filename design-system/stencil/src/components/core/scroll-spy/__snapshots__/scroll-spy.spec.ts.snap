// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooScrollSpyComponent Rendering Should render with a basic structure 1`] = `
<yoo-scroll-spy
  class="hydrated"
  data-yoo-scroll-spy-host=""
>
  
  <div
    data-yoo-scroll-spy=""
  />
</yoo-scroll-spy>
`;
